### PR TITLE
Introduce test runner biolerplate

### PR DESCRIPTION
**Pull request by seunlanlege**
companion of paritytech/substrate#9302

### CODE REVIEW

File: node/test/runtime/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+name = "polkadot-runtime-test"
+version = "0.8.28"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+
+test-runner = { git = "https://github.com/paritytech/substrate", branch = "substrate-test-runner" }
```
Comment (ordian) [inline node/test/runtime/Cargo.toml:11]: please make sure to run `cargo check` to update the `Cargo.lock` so that bot will be able to merge this properly
CI here will be red anyway, but the build should be checked on the substrate side 
```suggestion
test-runner = { git = "https://github.com/paritytech/substrate", branch = "master" }
```

**Comment by seunlanlege** [comment]
@ordian @kianenigma @bkchr  could you please take a look at this?

File: Cargo.lock [hunk 2]
```diff
@@ -9697,14 +9803,59 @@ dependencies = [
...
+version = "0.9.0"
+source = "git+https://github.com/paritytech/substrate?branch=master#2e34d473e6ba81b454eb039544f2162975c0b30b"
+dependencies = [
+ "env_logger 0.7.1",
+ "frame-system",
+ "futures 0.1.29",
+ "futures 0.3.13",
+ "jsonrpc-core",
+ "log",
+ "parity-scale-codec 1.3.7",
```
Comment (ordian) [inline Cargo.lock:9825]: why does it depend on parity-scale-codec v1 and not v2?

Comment (seunlanlege) [reply Cargo.lock:9812]: ahh, must be vestigial, can quickly bump it

Comment (ordian) [reply Cargo.lock:9812]: please address https://github.com/paritytech/substrate/pull/8485#discussion_r603587200 before you do

File: node/test/runtime/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+sp-keyring = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sp-api = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sp-inherents = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sp-keystore = { git = "https://github.com/paritytech/substrate", branch = "master" }
+
+polkadot-service = { path = "../../../node/service", version = "0.8.3" }
+polkadot-runtime = { path = "../../../runtime/polkadot", version = "0.8.28" }
+polkadot-runtime-common = { path = "../../../runtime/common", version = "0.8.28" }
+
+rand = "0.7"
```
Comment (ordian) [inline node/test/runtime/Cargo.toml:47]: another unused dependency?

**Comment by seunlanlege** [comment]
@ordian please could you look at this again?

File: node/test/runtime/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+grandpa = { package = "sc-finality-grandpa", git = "https://github.com/paritytech/substrate", branch = "master" }
+sp-consensus-babe = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-consensus-babe = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-consensus-manual-seal = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-service = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-executor = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-consensus = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-network = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-client-api = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sc-informant = { git = "https://github.com/paritytech/substrate", branch = "master" }
```
Comment (ordian) [inline node/test/runtime/Cargo.toml:35]: could you make sure there are no unused deps here?

**Comment by seunlanlege** [comment]
@kianenigma @ordian @gnunicorn Please drop a review so i can merge the companion pr

**Comment by ordian** [comment]
doesn't compile

**Comment by seunlanlege** [comment]
@ordian Yeah I think that's how it's supposed to be?, Until the companion goes green?

**Comment by ordian** [comment]
I'm lacking some context here, what's the relationship between `node/test/polkadot-simnet` and `simnet_tests`. There are some binaries there (`main.rs`), but are they integrated into CI? @drahnr do you know how it fits into the testing story?
https://github.com/paritytech/polkadot/blob/master/doc/testing.md

File: node/test/runtime/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+system = { package = "frame-system", git = "https://github.com/paritytech/substrate", branch = "master" }
+support = { package = "frame-support", git = "https://github.com/paritytech/substrate", branch = "master" }
+benchmarking = { package = "frame-benchmarking", git = "https://github.com/paritytech/substrate", branch = "master" }
+
+transaction-payment = { package = "pallet-transaction-payment", git = "https://github.com/paritytech/substrate", branch = "master" }
+collective = { package = "pallet-collective", git = "https://github.com/paritytech/substrate", branch = "master" }
+democracy = { package = "pallet-democracy", git = "https://github.com/paritytech/substrate", branch = "master" }
+
+polkadot-primitives = { version = "0.9", path = "../../../primitives" }
```
Comment (drahnr) [inline node/test/runtime/Cargo.toml:19]: Is there a reason we tie this to a version?

Comment (seunlanlege) [reply node/test/runtime/Cargo.toml:19]: not really, no

**Comment by seunlanlege** [comment]
> what's the relationship between node/test/polkadot-simnet and simnet_tests. 

The intention here is to run the test-runner as a binary, one binary is a simple node (`polkadot-simnet`), it receives blocks over the network, and accepts rpc requests.

https://github.com/paritytech/polkadot/blob/51789bf75f0f1273a2bba21b194f00ee6cb27dc0/node/test/polkadot-simnet/src/main.rs#L31-L52

The other binary (`simnet-runtime-upgrade`) actually runs the tests over the network, runtime-upgrades, migrations tests, what have you. 

https://github.com/paritytech/polkadot/blob/51789bf75f0f1273a2bba21b194f00ee6cb27dc0/node/test/polkadot-simnet/simnet-runtime-upgrade/src/main.rs#L25-L51

File: node/test/polkadot-simnet/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Binary used for simnet nodes, supports all runtimes, although only polkadot is implemented currently.
+//! This binary accepts all the cli args the polkadot binary does, Only difference is it uses
+//! manual-seal™ and babe for block authorship, it has a no-op verifier, so all blocks recieved over the network
+//! are imported and executed straight away. Block authorship/Finalization maybe done by calling the `engine_createBlock` &
+//! `engine_FinalizeBlock` rpc methods respectively.
```
Comment (ordian) [inline node/test/polkadot-simnet/src/main.rs:21]:
```suggestion
//! manual-seal™ and babe for block authorship, it has a no-op verifier, so all blocks received over the network
//! are imported and executed straight away. Block authorship/Finalization maybe done by calling the
//!  `engine_createBlock` & `engine_FinalizeBlock` rpc methods respectively.
```

File: node/test/polkadot-simnet/build-bin-remote.Dockerfile [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+RUN groupadd --gid 10001 nonroot && \
+    useradd  --home-dir /home/nonroot \
+             --create-home \
+             --shell /bin/bash \
+             --gid nonroot \
+             --groups nonroot \
+             --uid 10000 nonroot
+WORKDIR /home/nonroot/polkadot-simnet
+# Before running this cmd you need to manually copy polkadot-simnet binary from
+# project root target/reloease here, becasue all files in /target are ignored
```
Comment (ordian) [inline node/test/polkadot-simnet/build-bin-remote.Dockerfile:37]:
```suggestion
# project root target/release here, because all files in /target are ignored
```

File: node/test/polkadot-simnet/build-bin-remote.Dockerfile [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+# Use the non-root user to run our application
+# Tell run test script that it runs in container
+USER nonroot
+# check if executable works in this container
+RUN /usr/local/bin/polkadot-simnet --version
+# Tini allows us to avoid several Docker edge cases, see https://github.com/krallin/tini.
+ENTRYPOINT ["tini", "--", "/usr/local/bin/polkadot-simnet"]
+
+
```
Comment (ordian) [inline node/test/polkadot-simnet/build-bin-remote.Dockerfile:51]:
```suggestion
```

**Review by ordian — DISMISSED** [review]

**Comment by ordian** [comment]
> The intention here is to run the test-runner as a binary

so it's not integrated into CI yet, but will be incorporated into the simnet run check, is that correct?

**Comment by seunlanlege** [comment]
> so it's not integrated into CI yet, but will be incorporated into the simnet run check, is that correct?

yes once it's merged

File: node/test/polkadot-simnet/cheatsheet-commands.sh [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+# docker push paritypr/forrestgump:"$1"
+
+# build docker image locally using cache from cargo-chef
+# add export DOCKER_BUILDKIT=1  to ~/.bashrc
+
+# normally we use chargo-chef because it's caching capacity; add the --no-cache arg only
+# once in a while when the build fails
+date ; docker build -t seunlanlege/forrestgump:"$1" -f local-build-1-stages.Dockerfile --no-cache ../../../../
+
+date ; docker push seunlanlege/forrestgump:"$1"
```
Comment (drahnr) [inline node/test/polkadot-simnet/cheatsheet-commands.sh:17]: Not sure you want this commited, do you?

**Review by drahnr — CHANGES_REQUESTED** [review]
A very nice addition, but it needs some additional attention for how to use the docker files in practice, there are a few stray things that need cleanup, i.e. adding `#[deny(unused_crate_dependencies)]` would avoid unused deps @ordian pointed out.

Given that malus should be used in the same way your test boiler plate is supposed to be used, it mandates some consolidation on how generalize at least some part of it.

File: runtime/polkadot/src/lib.rs [hunk 10]
```diff
@@ -95,8 +95,8 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	spec_name: create_runtime_str!("polkadot"),
 	impl_name: create_runtime_str!("parity-polkadot"),
 	authoring_version: 0,
-	spec_version: 9090,
-	impl_version: 0,
+	spec_version: 9080,
+	impl_version: 1,
```
[the remark this answers is not in the document]
Comment (seunlanlege) [reply runtime/polkadot/src/lib.rs]: merge fallout, reverted

File: node/test/polkadot-simnet/cheatsheet-commands.sh (hunk 9 above)
Comment (seunlanlege) [reply node/test/polkadot-simnet/cheatsheet-commands.sh:17]: we're still working on the docker files, this will be removed before it is merged

**Comment by drahnr** [comment]
Could you also adapt/expand the `docs/testing.md` on how to use the `Dockerfile`s there (or a separate file in the same directory), that'd be much appreciated!

File: Cargo.lock [hunk 11]
```diff
@@ -47,9 +36,9 @@ dependencies = [
 
 [[package]]
 name = "aes"
-version = "0.4.0"
+version = "0.5.0"
```
Comment (ordian) [inline Cargo.lock:39]: please revert this `cargo update` again

File: node/test/polkadot-simnet/test/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+            })
+            .collect::<Vec<_>>();
+        // make sure transfer went through
+        assert_eq!(events.len(), 1);
+
+        // we're done, drop node.
+        drop(node);
+
+         Ok(())
+    })
```
Comment (ordian) [inline node/test/polkadot-simnet/test/src/main.rs:85]: spaces

**Review by gilescope — APPROVED** [review]

**Comment by ordian** [comment]
every new file uses spaces instead of tabs, please fix that as well as `cargo update` reversion

File: runtime/polkadot/src/lib.rs [hunk 13]
```diff
@@ -96,7 +96,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	impl_name: create_runtime_str!("parity-polkadot"),
 	authoring_version: 0,
 	spec_version: 9090,
-	impl_version: 0,
+	impl_version: 1,
```
Comment (ordian) [inline runtime/polkadot/src/lib.rs:99]:
```suggestion
	impl_version: 0,
```

**Comment by seunlanlege** [comment]
> every new file uses spaces instead of tabs, please fix that as well as cargo update reversion

damn, vscode strikes again

File: runtime/polkadot/src/lib.rs (hunk 13 above)
Comment (seunlanlege) [reply runtime/polkadot/src/lib.rs:99]: so this is actually required because of the changes on line 568 and 520

Comment (ordian) [reply runtime/polkadot/src/lib.rs:99]: the spec_version has been bumped already since the last release, so I don't so

File: node/test/polkadot-simnet/test/src/main.rs (hunk 12 above)
Comment (ordian) [reply node/test/polkadot-simnet/test/src/main.rs:85]: this file is still not tabified

Comment (ordian) [reply node/test/polkadot-simnet/test/src/main.rs:85]: unresolved this convo, because it's still not resolved

**Review by ordian — APPROVED** [review]

File: node/test/polkadot-simnet/common/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+		F: FnOnce(Node<PolkadotChainInfo>) -> Fut,
+		Fut: Future<Output=Result<(), Box<dyn Error>>>,
+{
+	use structopt::StructOpt;
+	use sc_cli::{CliConfiguration, SubstrateCli};
+
+	let mut tokio_runtime = build_runtime()?;
+	let task_executor = task_executor(tokio_runtime.handle().clone());
+	// parse cli args
+	let cmd = <polkadot_cli::Cli as StructOpt>::from_args();
```
Comment (drahnr) [inline node/test/polkadot-simnet/common/src/lib.rs:302]: There should be variant of this `run` that allows passing the `Cli` args explicitly rather than expecting them to present in the environment.

File: node/test/polkadot-simnet/test/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "polkadot-simnet-test"
+version = "0.1.0"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
+
+[dependencies]
+# substrate primitives
+sp-core = { git = "https://github.com/paritytech/substrate", branch = "master" }
```
Comment (drahnr) [inline node/test/polkadot-simnet/test/Cargo.toml:9]: @ordian iiuc this would cause the same issues that I accidentally introduced via #2962 and you resolved in https://github.com/paritytech/polkadot/pull/3443/commits/505b46c52939c770bd89db5b0eb9bf4f931ff35a

**Review by drahnr — APPROVED** [review]
<del>Recently I introduced some resolution issue for `sp-core` and iiuc this would re-introduce that problem, iff @ordian was aware of that and accounted for in his review,</del>LGTM

File: node/test/polkadot-simnet/test/Cargo.toml (hunk 15 above)
Comment (ordian) [reply node/test/polkadot-simnet/test/Cargo.toml:9]: I don't think so, that issue was related to `trybuild`, this is a separate test

Comment (drahnr) [reply node/test/polkadot-simnet/test/Cargo.toml:9]: My bad, ignore this then.

File: node/test/polkadot-simnet/common/src/lib.rs (hunk 14 above)
Comment (seunlanlege) [reply node/test/polkadot-simnet/common/src/lib.rs:302]: hmm, interesting but what would be the use case?

Comment (drahnr) [reply node/test/polkadot-simnet/common/src/lib.rs:302]: My idea was to also use this for small scale integration test level networks, launching 3 nodes i.e. which should be smth a regular dev laptop can handle.

Comment (seunlanlege) [reply node/test/polkadot-simnet/common/src/lib.rs:302]: could be added in a future PR, when the need for the functionality arises
